services/kafka/sasl.go

```diff
@@ -0,0 +1,108 @@
+package kafka
+
+import (
+	"errors"
+
+	sarama "github.com/Shopify/sarama"
+)
+
+type SASLAuth struct {
+	SASLUsername  string `toml:"sasl-username" override:"sasl-username"`
+	SASLPassword  string `toml:"sasl-password" override:"sasl-password"`
+	SASLMechanism string `toml:"sasl-mechanism" override:"sasl-mechanism"`
+	SASLVersion   *int   `toml:"sasl-version" override:"sasl-version"`
+
+	// GSSAPI config
+	SASLGSSAPIServiceName        string `toml:"sasl-gssapi-service-name" override:"sasl-gssapi-service-name"`
+	SASLGSSAPIAuthType           string `toml:"sasl-gssapi-auth-type" override:"sasl-gssapi-auth-type"`
+	SASLGSSAPIDisablePAFXFAST    bool   `toml:"sasl-gssapi-disable-pafxfast" override:"sasl-gssapi-disable-pafxfast"`
+	SASLGSSAPIKerberosConfigPath string `toml:"sasl-gssapi-kerberos-config-path" override:"sasl-gssapi-kerberos-config-path"`
+	SASLGSSAPIKeyTabPath         string `toml:"sasl-gssapi-key-tab-path" override:"sasl-gssapi-key-tab-path"`
+	SASLGSSAPIRealm              string `toml:"sasl-gssapi-realm" override:"sasl-gssapi-realm"`
+
+	// OAUTHBEARER config. experimental. undoubtedly this is not good enough.
+	SASLAccessToken string `toml:"sasl-access-token" override:"sasl-access-token"`
+}
+
+// SetSASLConfig configures SASL for kafka (sarama)
+func (k *SASLAuth) SetSASLConfig(config *sarama.Config) error {
```

Comment:
Rather than mutating, any chance we can just return the `config.Net.SASL` structure? That would make the call site more obvious.

Reply:
the config struct isn't clean when it is created, and defaults have already been populated in it by the constructor.